### DISCUSSION TITLE
Trouble rendering in debug mode on macOS.

### DISCUSSION THREAD

**Discussion by 5nefarious** [gfx-rs/wgpu#1437] [Q&A]
Hello,

I'm trying to write a test program in the [bevy](https://github.com/bevyengine/bevy) engine using wgpu-rs directly (bypassing the existing rendering infrastructure). So far, I've just copied the code from the `hello-triangle` example and integrated it into the engine. However, when I run the code in debug mode, I get a blank window. The triangle shows up just fine when I run with `--release`.

I can get the triangle to show up in debug mode if I print the `wgpu::Surface` (using `fmt::Debug`) right after it is created, but this seems like odd behavior. I cloned the wgpu-rs repo and was able to run the `hello-triangle` example successfully in both debug and release modes.

Here's the section of code that I suspect is relevant. I can post more if needed.
```rust
impl Renderer {
    pub async fn new(
        instance: &wgpu::Instance,
        window: &Window,
        winit_window: &winit::window::Window,
    ) -> Self {
        let window_id = window.id();
        let surface = unsafe { instance.create_surface(winit_window) };
        let adapter = instance
            .request_adapter(&wgpu::RequestAdapterOptions {
                power_preference: wgpu::PowerPreference::default(),
                compatible_surface: Some(&surface),
            })
            .await
            .expect("Failed to find an appropriate adapter");

        ...
```

I'm using the Metal backend on macOS 11.2.3. wgpu-rs is at version 0.7 (I can't seem to use `master` due to a dependency conflict).

**Comment by kvark** [comment]
Perhaps, you are dropping `Surface` early? We don't have the logic in place yet to hold it alive for you.

**Reply by 5nefarious** [reply]
I am saving the `Surface` in a `struct` that the engine holds on to.
```rust
pub struct Renderer {
    pub window_id: WindowId,
    surface: wgpu::Surface,
    device: wgpu::Device,
    queue: wgpu::Queue,
    swapchain_desc: wgpu::SwapChainDescriptor,
    swapchain: wgpu::SwapChain,
    render_pipeline: wgpu::RenderPipeline,
}
```
I don't see any reason why the `Surface` would be deallocated. How would this explain the inconsistent behavior between release and debug modes, though?

**Reply by kvark** [reply]
In Debug, freed memory can be filled with special markers, so trying to access a `Surface` that's deleted would cause reading garbage. In Release, the memory is just marked as free, but accessing it could still read the same data.

What about `Instance`? Are you dropping it to the floor too early?

**Reply by 5nefarious** [reply]
Thanks for the reply. I am holding onto `Instance` in a separate struct that is also stored by the engine. Also, it appears that the program doesn't consistently work in release mode, and it sometimes works in debug mode. It appears to be random.

**Reply by kvark** [reply]
One thing you could try is recording an API trace. See https://github.com/gfx-rs/wgpu/wiki/Debugging-wgpu-Applications#tracing-infrastructure
It would show what's wrong, potentially.
If not that, then we'd need to start looking for differences between your code and our examples.

**Reply by 5nefarious** [reply]
Okay, that trace tool was _super_ useful. I found my problem: wasn't holding on to the `ShaderModule`. Shaders were getting destroyed right after the renderer was constructed. Thanks for your help!

**Reply by kvark** [reply]
oh interesting. What kind of errors were you getting? We shouldn't need the shaders to be alive after you used them for pipeline creation.

**Reply by 5nefarious** [reply]
That's the thing—I wasn't getting _any_ errors. As far as `stdout` was concerned, the application was running just fine. It was only when I looked in the trace that I saw this gem:
```rust
DestroyShaderModule(Id(0, 1, Metal)),
```

In any case, adding a `shaders` field to my Renderer and saving the shader module to it seems to have fixed the issue.

**Reply by kvark** [reply]
It looks to still be an issue to me. There is nothing in a shader module that needs to be alive after you use it, and you can't use it if you already dropped it. So in no case you'd expect to run into a blank window.
What if you replay the API trace you recorded, does it also show as blank?

**Reply by 5nefarious** [reply]
I just tried running the player, and it throws the following error:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { code: Message("missing field `max_texture_dimension_1d`"), position: Position { line: 0, col: 0 } }', player/src/bin/play.rs:37:70
```

**Reply by kvark** [reply]
you need the `wgpu` of the same revision as used for recording

**Reply by 5nefarious** [reply]
Okay. So, I tried running the trace, and it seems to work just fine. The triangle shows up every time. Something is different about my code, then. I've pushed what I have so far to [GitHub](https://github.com/5nefarious/render-test), in case you have time to look at it.

I am currently trying to get a texture to display in the window, but I'm running into the same issue; It only works some of the time. I'm saving the `BindGroup`, but that doesn't seem to be cutting it.

**EDIT:** Similarly, when I run the player on the trace for my texture demo, the texture always shows up. Even though I got a blank window when I actually ran the application to produce that trace.

**Reply by 5nefarious** [reply]
For posterity, I opened up an issue here: #816.
